src/parse.rs

```diff
@@ -127,13 +127,42 @@ pub fn parse_fetches(
 pub fn parse_expunge(
     lines: Vec<u8>,
     unsolicited: &mut mpsc::Sender<UnsolicitedResponse>,
-) -> Result<Vec<u32>> {
-    let f = |resp| match resp {
-        Response::Expunge(id) => Ok(MapOrNot::Map(id)),
-        resp => Ok(MapOrNot::Not(resp)),
-    };
+) -> Result<Deleted> {
+    let mut lines: &[u8] = &lines;
+    let mut expunged = Vec::new();
+    let mut vanished = Vec::new();
+
+    loop {
+        if lines.is_empty() {
+            break;
+        }
+
+        match imap_proto::parser::parse_response(lines) {
+            Ok((rest, Response::Expunge(seq))) => {
+                lines = rest;
+                expunged.push(seq);
+            }
+            Ok((rest, Response::Vanished { earlier: _, uids })) => {
+                lines = rest;
+                vanished.extend(uids);
+            }
+            Ok((rest, data)) => {
+                lines = rest;
+                if let Some(resp) = handle_unilateral(data, unsolicited) {
+                    return Err(resp.into());
+                }
+            }
+            _ => {
+                return Err(Error::Parse(ParseError::Invalid(lines.to_vec())));
+            }
+        }
+    }
 
-    unsafe { parse_many(lines, f, unsolicited).map(|ids| ids.take()) }
+    if !vanished.is_empty() {
```

Comment:
Can there be both vanished and expunged results at once?

Reply:
According to the spec, no. Once the client has enabled QRESYNC then the server must send only VANISHED responses in lieu of EXPUNGE responses.

https://tools.ietf.org/html/rfc7162#section-3.2.10

Reply:
Ah, excellent! Could you please add a comment (with that link) pointing out that fact?

Reply:
Done!